### ISSUE TITLE
Net/SCP::get() adds a one-byte chr(0) at the end of the saved file

### ISSUE THREAD

**Issue by DavidAnderson684** [phpseclib/phpseclib#146]
In my testing, I just have a simple usage of Net_SCP::get(), with the second parameter ($local_file) set to true.

The resulting file is one byte larger than the file being fetched. Using "hexdump -C" shows that the difference is an extra chr(0) at the end of the file.

Testing shows that the server sends back the correct size in the line checked with preg_match() in get(). So, it looks to me like the while loop (while ($size < $info['size']) {) needs to make sure that it truncates $data such that the indicated amount of data isn't exceeded. (Presumably the protocol adds a the chr(0) - get() just needs to make sure it doesn't read that far).

On an entirely different matter - have you considered adding the FISH protocol? (http://en.wikipedia.org/wiki/Files_transferred_over_shell_protocol)

**Comment by aguidrevitch** [comment]
I have even worse situation:

[...@... uploads]$ ls -la 2014/06/Screen-Shot-2014-06-18-at-1.52.05-PM.png
-rw-rw-r-- 1 apache apache 638221 Jun 18  2014 2014/06/Screen-Shot-2014-06-18-at-1.52.05-PM.png

[...@... uploads]$ echo -e '\x00' |  scp -f 2014/06/Screen-Shot-2014-06-18-at-1.52.05-PM.png
C0664 638221 Screen-Shot-2014-06-18-at-1.52.05-PM.png

But when using phpseclib, $this->_send('\0') to scp -f 2014/06/Screen-Shot-2014-06-18-at-1.52.05-PM.png results in:
C0644 638222 Screen-Shot-2014-06-18-at-1.52.05-PM.png

And after it gets actually saved, size grows to 638223 (2 zeros in the end, md5sum of first 638221 is equal), with DavidAnderson684's patch, I have 638222

**Comment by vbresults** [comment]
@bantu @terrafrost Experiencing this with v2.0.6, PHP 7.1.1.  My 'client' is Linux, and the SSH 'server' is OSX.  Maybe there is some sort of protocol implementation difference between OSes?

**Comment by Jelle-S** [comment]
We are experiencing the same issue. Reading through this issue it seems like there is a possible fix for it that won't get implemented? Although I can't say I completely understand the reasoning why it won't.

**Comment by terrafrost** [comment]
> We are experiencing the same issue. Reading through this issue it seems like there is a possible fix for it that won't get implemented? Although I can't say I completely understand the reasoning why it won't.

Quoting an earlier post:

> I'm leaning towards not doing that actually. Net_SFTP doesn't do that for one and it's never posed any problems there. Net_SFTP does truncate downloaded material if you've explicitly told it the size it should be but if you didn't tell it the size it doesn't truncate it to match the reported file size.
>
> Among other things does scp lock files? If not then it's possible the initially reported file size could change if it was being modified as you were downloading it idk.
>
> Also, if download files were larger when downloaded than they were on the server it might be good to know. idk that throwing an error is justified but idk that trimming it to the reported length is the best approach either..

Also,

> Hiding problems with quick hacks that don't address the core of the problem makes identifying the core of the problem (let alone the fact that there is a problem) that much more difficult.

**Comment by Jelle-S** [comment]
I think I understand your concerns, but what is the way forward now? Is there a plan of attack? Will this get fixed in an other way? Is there anything I can do (besides making a PR of the code suggested above since that will most likely get declined for reasons stated in your comment above this one)?

**Comment by terrafrost** [comment]
> Is there a plan of attack?

No. It's just not a priority. I'm not planning on including SCP in the next version of phpseclib but I'm also not going to spend a significant amount of time to figure this out when there are much more important and useful features that phpseclib is in bad need of (I've been working on ECDSA / EdDSA support for phpseclib for a while and I'd rank that as a whole heck of a lot more important than this).

I mean, if you can give me SSH access to a server that reliably reproduces the problem that would probably help (you could email the auth info to terrafrost@php.net). But that, in and of itself, may not be sufficient. And even with that it's not going to be my priority. I might very well take a look this weekend if I get the itch to do so but my priorities are my priorities and I'm not prepared to change them.

**Comment by terrafrost** [comment]
I believe this should be fixed with https://github.com/phpseclib/phpseclib/commit/303a29c1933a406b30e1c6bcdeed45338982e151

**Comment by bantu** [comment]
@terrafrost Congrats for finding this!
